lib/src/time_util.rs

```diff
@@ -0,0 +1,189 @@
+use chrono_english::{parse_date_string, DateError, Dialect};
+use thiserror::Error;
+
+use crate::backend::Timestamp;
+
+/// Context needed to create a DatePattern during revset evaluation.
+#[derive(Clone, Debug)]
+pub enum DatePatternContext {
+    /// Interpret date patterns using the local machine's time zone
+    Local(DateTime<Local>),
+    /// Interpret date patterns using UTC
+    Utc(DateTime<Utc>),
+}
+
+impl DatePatternContext {
+    /// Parses a DatePattern from the given string and kind.
+    pub fn to_date_pattern(
+        &self,
+        s: &str,
+        kind: &str,
+    ) -> Result<DatePattern, DatePatternParseError> {
+        match &self {
+            DatePatternContext::Local(dt) => DatePattern::from_str_kind(s, kind, dt.to_owned()),
+            DatePatternContext::Utc(dt) => DatePattern::from_str_kind(s, kind, dt.to_owned()),
+        }
+    }
+}
+
+impl From<DateTime<Local>> for DatePatternContext {
+    fn from(value: DateTime<Local>) -> Self {
+        DatePatternContext::Local(value)
+    }
+}
+
+impl From<DateTime<Utc>> for DatePatternContext {
+    fn from(value: DateTime<Utc>) -> Self {
+        DatePatternContext::Utc(value)
+    }
+}
+
+/// Error occurred during date pattern parsing.
+#[derive(Debug, Error)]
+pub enum DatePatternParseError {
+    /// Unknown pattern kind is specified.
+    #[error(r#"Invalid date pattern kind "{0}:""#)]
+    InvalidKind(String),
+    /// Failed to parse timestamp.
+    #[error(transparent)]
+    ParseError(#[from] DateError),
+}
+
+/// Represents an range of dates that may be matched against.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DatePattern {
+    /// Represents all dates at or after the given instant.
+    AtOrAfter(Timestamp),
+    /// Represents all dates before, but not including, the given instant.
+    Before(Timestamp),
+}
+
+impl DatePattern {
+    /// Parses a string into a DatePattern.
+    ///
+    /// * `s` is the string to be parsed.
+    ///
+    /// * `kind` must be either "after" or "before". This determines whether the
+    ///   pattern will match dates after or before the parsed date.
+    ///
+    /// * `now` is the user's current time. This is a DateTime<Tz> because
+    ///   knowledge of offset changes is needed to correctly process relative
+    ///   times like "today". For example, California entered DST on March 10,
+    ///   2024, shifting clocks from UTC-8 to UTC-7 at 2:00 AM. If the pattern
+    ///   "today" was parsed at noon on that day, it should be interpreted as
+    ///   2024-03-10T00:00:00-08:00 even though the current offset is -07:00.
+    pub fn from_str_kind<Tz: TimeZone>(
+        s: &str,
+        kind: &str,
+        now: DateTime<Tz>,
+    ) -> Result<DatePattern, DatePatternParseError>
+    where
+        Tz::Offset: Copy,
+    {
+        let d =
+            parse_date_string(s, now, Dialect::Us).map_err(DatePatternParseError::ParseError)?;
+        let ts = Timestamp::from_datetime(d.to_utc().fixed_offset());
```

Comment:
Maybe `.to_utc()` is unneeded?

We can also change the underlying type to `MillisSinceEpoch` to clarify that the tz offset doesn't matter.